Userland/head.cpp

```diff
@@ -1,25 +1,87 @@
+#include <errno.h>
 #include <stdio.h>
 #include <stdlib.h>
+#include <LibCore/CArgsParser.h>
+
+int head(String filename, bool print_filename, int line_count);
 
 int main(int argc, char** argv)
 {
-    // FIXME: Allow setting via command-line argument.
+    CArgsParser args_parser("head");
+
+    args_parser.add_arg("n", "lines", "Number of lines to print (default 10)");
+    args_parser.add_arg("q", "Never print filenames");
+    args_parser.add_arg("v", "Always print filenames");
+
+    CArgsParserResult args = args_parser.parse(argc, (const char**)argv);
+
     int line_count = 10;
+    if (args.is_present("n")) {
+        line_count = strtol(args.get("n").characters(), NULL, 10);
+        if (errno == EINVAL) {
+            args_parser.print_usage();
+            return -1;
+        }
+    }
 
-    if (argc > 1) {
-        for (int i = 1; i < argc; ++i) {
-            if (argv[i][0] == '-') {
-                line_count = atoi(&argv[i][1]);
+    Vector<String> files = args.get_single_values();
+
+    bool print_filenames = files.size() > 1;
+
+    if (args.is_present("v")) {
+        print_filenames = true;
+    } else if (args.is_present("q")) {
+        print_filenames = false;
+    }
+
+    int rc = 0;
+
+    if (files.size() == 0) {
+        if (head("", print_filenames, line_count) != 0) {
+            rc = 1;
+        }
+    } else {
+        for (auto it = files.begin(); it != files.end(); ++it) {
+            if (head(*it, print_filenames, line_count) != 0) {
+                rc = 1;
             }
         }
     }
 
+    return rc;
+}
+
+int head(String filename, bool print_filename, int line_count)
+{
+    FILE* fd = nullptr;
+
+    if (filename == "" || filename == "-") {
+        fd = stdin;
+        filename = "standard input";
+    } else {
+        fd = fopen(filename.characters(), "r");
+        if (!fd) {
+            fprintf(stderr, "can't open %s for reading: %s\n", filename.characters(), strerror(errno));
+            return 1;
+        }
+    }
+
+    if (print_filename) {
+        printf("==> %s <==\n", filename.characters());
+    }
+
     for (int line = 0; line < line_count; ++line) {
         char buffer[BUFSIZ];
-        auto* str = fgets(buffer, sizeof(buffer), stdin);
+        auto* str = fgets(buffer, sizeof(buffer), fd);
         if (!str)
             break;
-        fputs(str, stdout);
+        printf("%s", str);
```

Comment:
I would keep the old version of this code. `fputs()` avoids having to parse the format string since it's just `"%s"` anyway.

Reply:
The reason I changed it to the `printf` call instead was to avoid the newline from `puts` doubling up with the newline from the source data. I've changed this to instead strip the trailing newline from the source string, and made this into a `puts` call. I didn't make it `fputs` because I can't think of any reason we'd want to output to anything other than stdout.